### ISSUE TITLE
`get_map` causes Memory leak

### ISSUE THREAD

**Issue by zxch3n** [y-crdt/y-crdt#208]
The following example causes a memory leak in the current main branch.

- `commit 649ba7679e9312f857641984f82f45f7c9de90e9`

```rust
  let doc = Doc::new();
  let mut txn = doc.transact();
  let map = txn.get_map("abc");
  map.insert(&mut txn, "xx", v);
```

So I checked the Rc count:

```rust
  let mut doc = Doc::new();
  dbg!(Rc::strong_count(&doc.store.0));
  let mut txn = doc.transact();
  dbg!(Rc::strong_count(&doc.store.0));
  let mut map = txn.get_map("abc");
  dbg!(Rc::strong_count(&doc.store.0));
  let m = doc.store.0.clone();
  drop(txn);
  drop(doc);
  drop(map);
  dbg!(Rc::strong_count(&m));
```

And the output is:

```log
[yrs/src/transaction.rs:797] Rc::strong_count(&doc.store.0) = 1
[yrs/src/transaction.rs:799] Rc::strong_count(&doc.store.0) = 2
[yrs/src/transaction.rs:801] Rc::strong_count(&doc.store.0) = 3
[yrs/src/transaction.rs:806] Rc::strong_count(&m) = 2
```

So the count that had been increased by creating a map was not reclaimed when dropping. Is this on purpose to avoid UAF?

**Comment by Horusiath** [comment]
This is problem indeed. Thanks for noticing. Basically there are two issues:

1. Shared types shouldn't outlive their document.
2. `Branch` should keep a weak reference to store rather than the strong one.

Re. 1 - I'm not yet sure about the good approach, since using a generic lifetime params (eg. `TextRef<'doc>`) sounds good on the one side and I wanted to try this approach already, but on the other it could wreck havoc once we go to using this in host languages (yawsm, y-py).

**Comment by zxch3n** [comment]
Yeah, I'm also worried about the interface in wasm, which makes lifetime unapproachable. Using a weak reference would introduce a new error to the user space that `Map` can only be used when `Doc` is alive, but I think it may be acceptable.

**Comment by Horusiath** [comment]
This one should be fixed in v0.13 which has been released today. Feel free to reopen this issue or create a new one if that's not the case.
